File: pine-java/src/main/java/page/liam/pine/Registry.java
package page.liam.pine;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.util.*;
import java.util.function.Supplier;

public class Registry {
    private static final ObjectMapper mapper = new ObjectMapper().enable(SerializationFeature.INDENT_OUTPUT);

    private static final Set<String> RESERVED_KEYS = new HashSet<>(Arrays.asList(
            "type_name", "$metadata", "$code_info", "skip", "recall", "sources",
            "debug", "row_dependency", "common_defaults", "item_defaults",
            "for_branch_control", "data_parallel"
    ));

    private static final Registry GLOBAL = new Registry();

    private final Map<String, OperatorEntry> operators = new LinkedHashMap<>();

    public static Registry global() { return GLOBAL; }

    public void register(OperatorSchema schema, Supplier<Operator> factory) {
        if (schema.name == null || schema.name.isEmpty()) {
            throw new PineErrors.RegistryError("", "Register called with empty name");
        }
        if (schema.type == null) {
            throw new PineErrors.RegistryError(schema.name, "Type is required");
        }
        if (schema.description == null || schema.description.isEmpty()) {
            throw new PineErrors.RegistryError(schema.name, "Description is required");
        }
        for (Map.Entry<String, ParamSpec> entry : schema.params.entrySet()) {
            if (entry.getValue().description == null || entry.getValue().description.isEmpty()) {
                throw new PineErrors.RegistryError(schema.name,
                        "param \"" + entry.getKey() + "\": Description is required");
            }
        }
        if (operators.containsKey(schema.name)) {
            throw new PineErrors.RegistryError(schema.name, "already registered");
        }
        operators.put(schema.name, new OperatorEntry(schema, factory));
    }

    public Operator buildOperator(String typeName, Map<String, Object> rawParams) {
        OperatorEntry entry = operators.get(typeName);
        if (entry == null) {
            throw new PineErrors.RegistryError(typeName, "operator type not registered");
        }

        Map<String, Object> params = validateAndExtractParams(entry.schema, rawParams);

        Operator op = entry.factory.get();
        try {
            op.init(new OperatorParams(params));
        } catch (RuntimeException e) {
            throw new PineErrors.RegistryError(typeName, "Init failed: " + e.getMessage());
        }
        return op;
    }

    public OperatorType getType(String typeName) {
        OperatorEntry entry = operators.get(typeName);
        return entry != null ? entry.schema.type : null;
    }

    public OperatorSchema getSchema(String typeName) {
        OperatorEntry entry = operators.get(typeName);
        return entry != null ? entry.schema : null;
    }

    public List<OperatorSchema> all() {
        List<OperatorSchema> result = new ArrayList<>();
        for (OperatorEntry entry : operators.values()) {
            result.add(entry.schema);
        }
        result.sort(Comparator.comparing(s -> s.name));
        return result;
    }

    public String exportSchemaJSON() throws Exception {
        List<Map<String, Object>> schemas = new ArrayList<>();
        for (OperatorSchema s : all()) {
            Map<String, Object> obj = new LinkedHashMap<>();
            obj.put("Name", s.name);
            String typeName = s.type.name().charAt(0) + s.type.name().substring(1).toLowerCase();
            obj.put("Type", typeName);
            obj.put("Description", s.description);
            Map<String, Object> params = new LinkedHashMap<>();
            for (Map.Entry<String, ParamSpec> p : s.params.entrySet()) {
                Map<String, Object> spec = new LinkedHashMap<>();
                spec.put("Type", p.getValue().type);
                spec.put("Required", p.getValue().required);
                spec.put("Default", p.getValue().defaultValue);
                spec.put("Description", p.getValue().description);
                params.put(p.getKey(), spec);
            }
            obj.put("Params", params);
            schemas.add(obj);
        }
        return mapper.writeValueAsString(schemas);
    }

    public void reset() {
        operators.clear();
    }

    // --- Static convenience methods delegating to global instance ---

    public static void registerGlobal(OperatorSchema schema, Supplier<Operator> factory) {
        GLOBAL.register(schema, factory);
    }

    public static Map<String, Object> validateAndExtractParams(
            OperatorSchema schema, Map<String, Object> rawParams) {
        Map<String, Object> params = new LinkedHashMap<>();

        for (Map.Entry<String, Object> entry : rawParams.entrySet()) {
            if (!RESERVED_KEYS.contains(entry.getKey())) {
                params.put(entry.getKey(), entry.getValue());
            }
        }

        for (Map.Entry<String, ParamSpec> entry : schema.params.entrySet()) {
            String name = entry.getKey();
            ParamSpec spec = entry.getValue();
            if (!params.containsKey(name)) {
                if (spec.required) {
                    throw new PineErrors.RegistryError(schema.name,
                            "required parameter \"" + name + "\" missing for operator \"" + schema.name + "\"");
                }
                if (spec.defaultValue != null) {
                    params.put(name, spec.defaultValue);
                }
            }
        }

        for (String key : params.keySet()) {
            if (!schema.params.containsKey(key)) {
                throw new PineErrors.RegistryError(schema.name,
                        "unknown parameter \"" + key + "\" for operator \"" + schema.name + "\"");
            }
        }

        return params;
    }

    private static class OperatorEntry {
        final OperatorSchema schema;
        final Supplier<Operator> factory;

        OperatorEntry(OperatorSchema schema, Supplier<Operator> factory) {
            this.schema = schema;
            this.factory = factory;
        }
    }
}

File: pine-python/pine/registry.py
from __future__ import annotations

from typing import Any, Callable

from pine.config import RESERVED_KEYS  # noqa: F401 — re-export for backwards compat
from pine.errors import RegistryError
from pine.operator import Operator, OperatorParams, OperatorSchema


class Registry:
    """Operator type registry mapping type names to schemas and factories."""

    _global: "Registry | None" = None

    def __init__(self):
        self._schemas: dict[str, OperatorSchema] = {}
        self._factories: dict[str, Callable[[], Operator]] = {}

    @classmethod
    def global_instance(cls) -> "Registry":
        assert cls._global is not None, "Registry not initialized (circular import?)"
        return cls._global

    @classmethod
    def register_global(cls, schema: OperatorSchema, factory: Callable[[], Operator]):
        cls.global_instance().register(schema, factory)

    def register(self, schema: OperatorSchema, factory: Callable[[], Operator]):
        if not schema.name:
            raise RegistryError("operator schema name is empty")
        if not schema.description:
            raise RegistryError(f"operator \"{schema.name}\": description is empty")
        if schema.name in self._schemas:
            raise RegistryError(f"operator \"{schema.name}\" already registered")
        self._schemas[schema.name] = schema
        self._factories[schema.name] = factory

    def build_operator(self, type_name: str, raw_params: dict[str, Any]) -> Operator:
        schema = self._schemas.get(type_name)
        if schema is None:
            raise RegistryError(f"operator type not registered: \"{type_name}\"")
        factory = self._factories[type_name]

        params = self._validate_and_extract_params(schema, raw_params)
        op = factory()
        try:
            op.init(OperatorParams(params))
        except RegistryError:
            raise
        except Exception as e:
            raise RegistryError(f'operator "{type_name}": init failed: {e}') from e
        return op

    def _validate_and_extract_params(
        self, schema: OperatorSchema, raw_params: dict[str, Any]
    ) -> dict[str, Any]:
        params: dict[str, Any] = {}
        for key, value in raw_params.items():
            if key in RESERVED_KEYS:
                continue
            if key not in schema.params:
                raise RegistryError(
                    f"unknown parameter \"{key}\" for operator \"{schema.name}\""
                )
            params[key] = value

        for name, spec in schema.params.items():
            if name not in params:
                if spec.required:
                    raise RegistryError(
                        f"required parameter \"{name}\" missing for operator \"{schema.name}\""
                    )
                if spec.default_value is not None:
                    params[name] = spec.default_value

        return params

    def schemas(self) -> list[OperatorSchema]:
        return list(self._schemas.values())

    def has(self, type_name: str) -> bool:
        return type_name in self._schemas

    def get_schema(self, type_name: str) -> "OperatorSchema | None":
        return self._schemas.get(type_name)


Registry._global = Registry()
